### PR TITLE
Compilation time, breaking out storage_defs and hash_util

### PR DESCRIPTION
**Pull request by lmwnshn**
Another quick attempt at #1037. At a high level, hash_util (and transitively storage_defs) became expensive includes. So breaking those out of header files into cpp files is good.

- Measured with the Python script linked in #1037, [diff.csv.txt](https://github.com/cmu-db/terrier/files/5065020/diff.csv.txt)
  - hash_util.h : -222 includes
  - xxh3.h : -217 includes

- New build time, 23.13 minutes user time,
  - ```
    [100%] Built target terrier
    make -j12 terrier  1387.90s user 79.82s system 867% cpu 2:49.18 total
    ```
- We had hash_t defined in hash_util and strong_typedef. Shot the former.

### CODE REVIEW

File: src/execution/sql/runtime_types.cpp [hunk 1]
```diff
@@ -655,4 +661,18 @@ Timestamp Timestamp::FromYMDHMSMU(int32_t year, int32_t month, int32_t day, int3
   return Timestamp(result);
 }
 
+hash_t Timestamp::Hash(const hash_t seed) const { return common::HashUtil::HashCrc(value_, seed); }
```
Comment (lmwnshn) [inline src/execution/sql/runtime_types.cpp:664]: Header

**Comment by lmwnshn** [comment]
With latest push, still around 23 minutes.

```
[100%] Built target terrier
make -j12 terrier  1394.00s user 80.02s system 817% cpu 3:00.35 total
```

**Comment by mbutrovich** [comment]
I'm curious about those benchmark results on the Varlen stuff. I'll run it locally first to see how it looks.

**Comment by mbutrovich** [comment]
VarlenEntryBenchmark hashing is looking rough. We might need to break that class into its own header so we can keep stuff inlined but reduce the blast radius of the includes.

**Comment by mbutrovich** [comment]
master:
```
VarlenEntryBenchmark/HashInline/1                                          2 ns          2 ns  288496671   399.084MB/s   399.084M items/s
VarlenEntryBenchmark/HashInline/2                                          2 ns          2 ns  453828438   1.22606GB/s   627.743M items/s
VarlenEntryBenchmark/HashInline/4                                          2 ns          2 ns  310293258   1.65137GB/s    422.75M items/s
VarlenEntryBenchmark/HashInline/8                                          2 ns          2 ns  310302737   3.30279GB/s   422.757M items/s
VarlenEntryBenchmark/HashInline/12                                         3 ns          3 ns  232727795   3.71559GB/s   317.064M items/s
VarlenEntryBenchmark/HashNotInline/16                                      3 ns          3 ns  279266062     5.945GB/s    380.48M items/s
VarlenEntryBenchmark/HashNotInline/32                                      4 ns          4 ns  199481861   8.49285GB/s   271.771M items/s
VarlenEntryBenchmark/HashNotInline/64                                      5 ns          5 ns  132987414   11.3239GB/s   181.182M items/s
VarlenEntryBenchmark/HashNotInline/128                                    10 ns         10 ns   71609218   12.1948GB/s   97.5583M items/s
VarlenEntryBenchmark/HashNotInline/256                                    16 ns         16 ns   44353399   15.2644GB/s   61.0576M items/s
VarlenEntryBenchmark/HashNotInline/512                                    22 ns         22 ns   32142366   21.8914GB/s   43.7828M items/s
VarlenEntryBenchmark/HashNotInline/1024                                   46 ns         46 ns   15355544   20.9095GB/s   20.9095M items/s
VarlenEntryBenchmark/HashNotInline/2048                                   60 ns         60 ns   11566715   31.9334GB/s   15.9667M items/s
VarlenEntryBenchmark/HashNotInline/4096                                  116 ns        116 ns    6003818   32.8824GB/s   8.22061M items/s
```
#1074:
```
VarlenEntryBenchmark/HashInline/1                                          3 ns          3 ns  242296011   333.444MB/s   333.444M items/s
VarlenEntryBenchmark/HashInline/2                                          3 ns          3 ns  247005883   691.737MB/s   345.868M items/s
VarlenEntryBenchmark/HashInline/4                                          4 ns          4 ns  190526079   1028.13MB/s   257.033M items/s
VarlenEntryBenchmark/HashInline/8                                          3 ns          3 ns  279271580   2.96651GB/s   379.713M items/s
VarlenEntryBenchmark/HashInline/12                                         4 ns          4 ns  186181266   2.97243GB/s   253.648M items/s
VarlenEntryBenchmark/HashNotInline/16                                      3 ns          3 ns  209829459   4.50825GB/s   288.528M items/s
VarlenEntryBenchmark/HashNotInline/32                                      5 ns          5 ns  146153662   6.17626GB/s    197.64M items/s
VarlenEntryBenchmark/HashNotInline/64                                      7 ns          7 ns  103434295   8.79838GB/s   140.774M items/s
VarlenEntryBenchmark/HashNotInline/128                                    11 ns         11 ns   61212813   10.4382GB/s   83.5054M items/s
VarlenEntryBenchmark/HashNotInline/256                                    16 ns         16 ns   44671461   15.1102GB/s   60.4409M items/s
VarlenEntryBenchmark/HashNotInline/512                                    23 ns         23 ns   31037884   21.1606GB/s   42.3213M items/s
VarlenEntryBenchmark/HashNotInline/1024                                   46 ns         46 ns   15095750    20.568GB/s    20.568M items/s
VarlenEntryBenchmark/HashNotInline/2048                                   60 ns         60 ns   11638636   31.6105GB/s   15.8052M items/s
VarlenEntryBenchmark/HashNotInline/4096                                  116 ns        116 ns    6055534   32.8656GB/s   8.21641M items/s
```

**Comment by gonzalezjo** [comment]
Build fails for me with clang-8 (release mode) on Ubuntu 18.04 LTS: 
```
[100%] Linking CXX executable ../../release/varlen_entry_benchmark
../../release/libterrier.a(abstract_expression.cpp.o): In function `terrier::parser::AbstractExpression::Hash() const':
abstract_expression.cpp:(.text+0x262): undefined reference to `terrier::common::HashUtil::HashXX3(unsigned char const*, unsigned int)'
abstract_expression.cpp:(.text+0x2cb): undefined reference to `terrier::common::HashUtil::HashXX3(unsigned char const*, unsigned int)'
abstract_expression.cpp:(.text+0x2fe): undefined reference to `terrier::common::HashUtil::HashXX3(unsigned char const*, unsigned int)'
abstract_expression.cpp:(.text+0x331): undefined reference to `terrier::common::HashUtil::HashXX3(unsigned char const*, unsigned int)'
../../release/libterrier.a(constant_value_expression.cpp.o): In function `terrier::parser::ConstantValueExpression::Hash() const':
constant_value_expression.cpp:(.text+0xa35): undefined reference to `terrier::common::HashUtil::HashXX3(unsigned char const*, unsigned int)'
../../release/libterrier.a(select_statement.cpp.o):select_statement.cpp:(.text+0x69): more undefined references to `terrier::common::HashUtil::HashXX3(unsigned char const*, unsigned int)' follow
../../release/libterrier.a(storage_defs.cpp.o): In function `terrier::storage::VarlenEntry::Hash(unsigned long) const':
storage_defs.cpp:(.text+0xff): undefined reference to `terrier::common::HashUtil::HashXX3(unsigned char const*, unsigned int, unsigned long)'
../../release/libterrier.a(column_value_expression.cpp.o): In function `terrier::parser::ColumnValueExpression::Hash() const':
column_value_expression.cpp:(.text+0x352): undefined reference to `terrier::common::HashUtil::HashXX3(unsigned char const*, unsigned int)'
column_value_expression.cpp:(.text+0x36c): undefined reference to `terrier::common::HashUtil::HashXX3(unsigned char const*, unsigned int)'
column_value_expression.cpp:(.text+0x3ac): undefined reference to `terrier::common::HashUtil::HashXX3(unsigned char const*, unsigned int)'
column_value_expression.cpp:(.text+0x3e5): undefined reference to `terrier::common::HashUtil::HashXX3(unsigned char const*, unsigned int)'
column_value_expression.cpp:(.text+0x41c): undefined reference to `terrier::common::HashUtil::HashXX3(unsigned char const*, unsigned int)'
../../release/libterrier.a(column_value_expression.cpp.o):column_value_expression.cpp:(.text+0x453): more undefined references to `terrier::common::HashUtil::HashXX3(unsigned char const*, unsigned int)' follow
clang: error: linker command failed with exit code 1 (use -v to see invocation)
benchmark/storage/CMakeFiles/varlen_entry_benchmark.dir/build.make:158: recipe for target 'release/varlen_entry_benchmark' failed
make[3]: *** [release/varlen_entry_benchmark] Error 1
CMakeFiles/Makefile2:6348: recipe for target 'benchmark/storage/CMakeFiles/varlen_entry_benchmark.dir/all' failed
make[2]: *** [benchmark/storage/CMakeFiles/varlen_entry_benchmark.dir/all] Error 2
CMakeFiles/Makefile2:6355: recipe for target 'benchmark/storage/CMakeFiles/varlen_entry_benchmark.dir/rule' failed
make[1]: *** [benchmark/storage/CMakeFiles/varlen_entry_benchmark.dir/rule] Error 2
Makefile:2340: recipe for target 'varlen_entry_benchmark' failed
make: *** [varlen_entry_benchmark] Error 2
```

**Comment by lmwnshn** [comment]
The regression @mbutrovich points out is real and I can reproduce it by re-including xxh3 in hash_util (at significant cost to compile time). But I'm a little baffled.

The benchmark which gets slower is ostensibly for inlined varlen entries. That code isn't being touched at all.

**Comment by lmwnshn** [comment]
In other news, with the latest refactor we're down to 22 minutes on my box.

**Comment by gonzalezjo** [comment]
You can derive much of the benefit here just by simplifying xxhash. (See: https://github.com/gonzalezjo/terrier/pull/new/smallxxhash)

No performance regression from the latter.

**Comment by lmwnshn** [comment]
With #1152 in, maybe we don't need to be so invasive about breaking things up. Good beginner issue if anyone wants to pick it up, otherwise I'll poke at it again in the future when we bump toolchain.

**Pull request closed without merging**